### PR TITLE
Martial Mastery: Defensive Perks

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary
Mods "Martial Mastery: Add three new defensive perks."

#### Purpose of change
"So you know a thing or two about throwing a punch but what about taking them?"

#### Describe the solution

There's three new perks here:

- Shrug it off: Completely blocks to the first damage instance that hits you every 15 turns, whatever it is.
- Like water: Gives you a flat 25% chance to dodge melee attacks if you have at least 10 insight.
- Ballistic Evasion: Gives you a DEX x 1% chance to dodge projectiles, if you have moved at least 4 tiles the chance increases to DEX x 5%

#### Testing
Some combat.

### CODE REVIEW

File: data/mods/Perk_melee/PERKS/momentum.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "name": { "str": "Momentum II" },
+    "description": "Rapid movement generates momentum stacks.  As long as momentum is active your attacks gain extra physical damage equal to 10% of your dexterity.  You gain a point of dexterity for every 2 momentum stacks.",
+    "flags": "MELEE_PERK_MOMENTUM"
+  },
+  {
+    "type": "mutation",
+    "id": "MELEE_PERK_MOMENTUM_BULLET_DODGE",
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Ballistic Evasion" },
+    "description": "If you have at least 4 momentum stacks, gain a 5% chance to dodge any ranged attack for every point of dexterity you have. If you have less than 4 momentum stacks, gain a 1% chance to dodge any ranged attack for every point of dexterity you have.",
```
Comment (John-Candlebury) [inline data/mods/Perk_melee/PERKS/momentum.json:24]:
```suggestion
    "description": "If you have at least 4 momentum stacks, gain a 5% chance to dodge any ranged attack for every point of dexterity you have.  If you have less than 4 momentum stacks, gain a 1% chance to dodge any ranged attack for every point of dexterity you have.",
```

File: data/mods/Perk_melee/EOC/shrug_off.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    "id": "melee_perk_ench_shrug_off",
+    "condition": { "math": [ "u_shrug_off_ready", "==", "1" ] },
+    "name": { "str": "Shrug it off" },
+    "description": "You will completely negate the next attack that hits you.",
+    "values": [ { "value": "FORCEFIELD", "add": 1 } ],
+    "hit_me_effect": [
+      {
+        "id": "sp_shrug_off_hit",
+        "hit_self": true,
+        "message": "You shruf off the attack.",
```
Comment (John-Candlebury) [inline data/mods/Perk_melee/EOC/shrug_off.json:21]:
```suggestion
        "message": "You shrug off the attack.",
```

File: data/mods/Perk_melee/PERKS/insight.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    "name": { "str": "Sixth Sense" },
+    "description": "You accumulate up to 20 insight stacks even from enemies you can't see.  You can perceive unseen enemies within 10 tiles if you have at least 10 insight stacks.",
+    "flags": "MELEE_PERK_SIXTH_SENSE",
+    "enchantments": [ "melee_perk_ench_six_sense" ]
+  },
+  {
+    "type": "mutation",
+    "id": "MELEE_PERK_INSIGHT_INFINITE",
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "The psychic sea" },
```
Comment (Night-Pryanik) [inline data/mods/Perk_melee/PERKS/insight.json:33]: For consistency with other perk names.
```suggestion
    "name": { "str": "The Psychic Sea" },
```

File: data/mods/Perk_melee/PERKS/perks.json [hunk 4]
```diff
@@ -0,0 +1,76 @@
+[
+  {
+    "type": "mutation_type",
+    "id": "perk"
+  },
```
Comment (Night-Pryanik) [inline data/mods/Perk_melee/PERKS/perks.json:5]: Is this needed?

File: data/mods/Perk_melee/PERKS/perks.json [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Carefree Stance" },
+    "description": "You aren't even trying.\nYou clear all combat buffs whenever you wait.",
+    "flags": "MELEE_PERK_CAREFREE_STANCE"
+  },
+  {
+    "type": "mutation",
+    "id": "MELEE_PERK_MOVING_DODGE",
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Flowing retreat" },
```
Comment (Night-Pryanik) [inline data/mods/Perk_melee/PERKS/perks.json:37]:
```suggestion
    "name": { "str": "Flowing Retreat" },
```

File: data/mods/Perk_melee/PERKS/tempo.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Tempo Shift" },
+    "description": "Gain a stack of combat tempo whenever you successfully block an attack.",
+    "flags": "MELEE_PERK_TEMPO_SHIFT"
+  },
+  {
+    "type": "mutation",
+    "id": "MELEE_PERK_TEMPO_THROW",
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Measured throws" },
```
Comment (Night-Pryanik) [inline data/mods/Perk_melee/PERKS/tempo.json:22]:
```suggestion
    "name": { "str": "Measured Throws" },
```

File: data/mods/Perk_melee/PERKS/insight.json [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    "flags": "MELEE_PERK_INSIGHT",
+    "enchantments": [ "melee_perk_ench_insight" ]
+  },
+  {
+    "type": "mutation",
+    "id": "MELEE_PERK_INSIGHT_DODGE",
+    "copy-from": "MELEE_PERK_BASE",
+    "name": { "str": "Like Water" },
+    "description": "You gain a +25 dodge chance if you have at least 10 insight stacks.",
+    "flags": "MELEE_PERK_LIKE_WATER",
```
Comment (Night-Pryanik) [inline data/mods/Perk_melee/PERKS/insight.json:17]: This flag seems to be unused?

Comment (John-Candlebury) [reply data/mods/Perk_melee/PERKS/insight.json:17]: I keep all these flags for cross modability or for future items having the techniques incorporated even if you dont have the perks.

Most are unused atm,

File: data/mods/Perk_melee/PERKS/perks.json (hunk 4 above)
Comment (John-Candlebury) [reply data/mods/Perk_melee/PERKS/perks.json:5]: I think its required for purifier reasons IRC its something bombasticperks found out.

**Pull request merged**